exporter/awsxrayexporter/conn.go

```diff
@@ -0,0 +1,276 @@
+	// Make explicit call to fetch credentials.
+	_, err := stsCred.Get()
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case sts.ErrCodeRegionDisabledException:
+				logger.Error("Region : %v - %v", zap.String("region", region), zap.String("error", aerr.Error()))
+				logger.Info("Credentials for provided RoleARN will be fetched from STS primary region endpoint instead of regional endpoint.")
+				stsCred = getSTSCredsFromPrimaryRegionEndpoint(logger, t, roleArn, region)
+			}
+		}
+	}
+	return stsCred
+}
+
+// getSTSCredsFromRegionEndpoint fetches STS credentials for provided roleARN from regional endpoint.
+// AWS STS recommends that you provide both the Region and endpoint when you make calls to a Regional endpoint.
+// Reference: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html#id_credentials_temp_enable-regions_writing_code
+func getSTSCredsFromRegionEndpoint(logger *zap.Logger, sess *session.Session, region string,
+	roleArn string) *credentials.Credentials {
+	regionalEndpoint := getSTSRegionalEndpoint(region)
+	// if regionalEndpoint is "", the STS endpoint is Global endpoint for classic regions except ap-east-1 - (HKG)
+	// for other opt-in regions, region value will create STS regional endpoint.
+	// This will be only in the case, if provided region is not present in aws_regions.go
+	c := &aws.Config{Region: aws.String(region), Endpoint: &regionalEndpoint}
+	st := sts.New(sess, c)
+	logger.Info("STS Endpoint : %v", zap.String("endpoint", st.Endpoint))
+	return stscreds.NewCredentialsWithClient(st, roleArn)
+}
+
+// getSTSCredsFromPrimaryRegionEndpoint fetches STS credentials for provided roleARN from primary region endpoint in
+// the respective partition.
+func getSTSCredsFromPrimaryRegionEndpoint(logger *zap.Logger, t *session.Session, roleArn string,
+	region string) *credentials.Credentials {
+	partitionID := getPartition(region)
+	if partitionID == endpoints.AwsPartitionID {
+		return getSTSCredsFromRegionEndpoint(logger, t, endpoints.UsEast1RegionID, roleArn)
+	} else if partitionID == endpoints.AwsCnPartitionID {
+		return getSTSCredsFromRegionEndpoint(logger, t, endpoints.CnNorth1RegionID, roleArn)
+	} else if partitionID == endpoints.AwsUsGovPartitionID {
+		return getSTSCredsFromRegionEndpoint(logger, t, endpoints.UsGovWest1RegionID, roleArn)
+	}
+
+	return nil
+}
+
+func getSTSRegionalEndpoint(r string) string {
+	p := getPartition(r)
+
+	var e string
+	if p == endpoints.AwsPartitionID || p == endpoints.AwsUsGovPartitionID {
+		e = STSEndpointPrefix + r + STSEndpointSuffix
+	} else if p == endpoints.AwsCnPartitionID {
+		e = STSEndpointPrefix + r + STSAwsCnPartitionIDSuffix
+	}
+	return e
+}
+
+func getDefaultSession(logger *zap.Logger) *session.Session {
+	result, serr := session.NewSession()
+	if serr != nil {
+		logger.Error("Error in creating session object : %v\n.", zap.Error(serr))
+		os.Exit(1)
```

Comment:
We should not be exiting on errors. Instead log if necessary and propagate the error to the caller.

Generally as a rule of thumb: Collector is a long-running process that needs to be autonomously robust. Exiting the process on an error will result in losing all data queued inside collector (if this function is called while the Collector is running).

Reply:
Fixed